apps/firebase-api/tests/users.courses.submissions.spec.ts

```diff
@@ -0,0 +1,178 @@
+import * as firebase from '@firebase/rules-unit-testing';
+
+import {
+  PROJECT_ID,
+  getAuthedFirestore,
+  updateUser,
+  getUserRef,
+} from './utils';
+
+import { getUserSubmissionRef } from './utils';
+
+const ALICE_ID = 'alice';
+const BOB_ID = 'bob';
+const DAN_ID = 'dan';
+const RANDOM_SUBMISSION_ID = 'random_review';
+const COURSE_A = 'course_a';
+const COURSE_B = 'course_b';
+
+describe('users/{{userID}}/courses/{{courseId}}/submissions/{{submissionId}}', () => {
+  beforeEach(async () => {
+    await firebase.clearFirestoreData({ projectId: PROJECT_ID });
+  });
+
+  it('create', async () => {
+    // bob is mentor of course_a
+    await updateUser(BOB_ID, {
+      is_mentor: true,
+      mentorable_courses: [COURSE_A],
+    });
+
+    // alice can create
+    const aliceDb = getAuthedFirestore({ uid: ALICE_ID });
+    await firebase.assertSucceeds(
+      getUserSubmissionRef(aliceDb, ALICE_ID, COURSE_A, RANDOM_SUBMISSION_ID).set({ new: true })
+    );
+
+    // others cannot create
+    const anyoneDb = getAuthedFirestore(null);
+    await firebase.assertFails(
+      getUserSubmissionRef(anyoneDb, ALICE_ID, COURSE_A, RANDOM_SUBMISSION_ID).set({ new: false })
+    );
+
+    // bob cannot create
+    const bobDb = getAuthedFirestore({ uid: BOB_ID });
+    await firebase.assertFails(
+      getUserSubmissionRef(bobDb, ALICE_ID, COURSE_A, RANDOM_SUBMISSION_ID).set({ new: false })
+    );
+  });
+
+  it('read: mentors or owner can read', async () => {
+    // bob is mentor of course_a
+    await updateUser(BOB_ID, {
+      is_mentor: true,
+      mentorable_courses: [COURSE_A],
+    });
+    // dan is mentor of course_b
+    await updateUser(DAN_ID, {
+      is_mentor: true,
+      mentorable_courses: [COURSE_B],
+    });
+
+    // alice can get
+    const aliceDb = getAuthedFirestore({ uid: ALICE_ID });
+    await firebase.assertSucceeds(
+      getUserSubmissionRef(aliceDb, ALICE_ID, COURSE_A, RANDOM_SUBMISSION_ID).set({
+        new: true,
+        mentor: getUserRef(aliceDb, BOB_ID),
```

Comment:
What is this for? You're trying to update a submission? Students should not be able to do that.

Reply:
Nor should they be able to set the `mentor` field. That's only set by the Firebase admin in functions. Nor is there a `new` field.